File: scripts/python/model_training.py
"""Core engine for training the weighted logistic regression classifier: protein-stratified
train-test splitting, model training, and prediction extraction.

Python port of scripts/model_training.R. The R version trains via caret::train(), which
also runs an inner 5-fold cross-validation to report Accuracy/ROC/Sens/Spec (see
accROCSummary()/fitControl_CV in the R script) - but method="glm" has no hyperparameters to
tune, so that inner CV never shapes or selects the final model; it only populates
mod$resample/mod$results fields that trainAndMakePreds_logreg() never reads. This port
skips reproducing that inner-CV machinery entirely and fits scikit-learn's
LogisticRegression directly - the outer protein-stratified test-set evaluation (auc_df /
test_preds_df), which IS what gets used, is unaffected.

For the weighted fit itself: R's glm(weights = w) maximises the weighted log-likelihood
sum(w_i * loglik_i) with no regularisation. The closest scikit-learn equivalent is
LogisticRegression(C=float("inf")).fit(X, y, sample_weight=w) - unregularised weighted MLE,
optimised with LBFGS instead of R's IRLS. Both are exact optimisers of the same convex
objective, so coefficients should match closely (small differences only from convergence
tolerance), even though the two frameworks disagree on standard errors/inference - which
this pipeline never uses anyway.
"""

import pandas as pd
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import roc_auc_score
from sklearn.model_selection import StratifiedKFold


##### Class weights #####

def case_weights_balanced(resp):
    """Per-sample weights giving the positive ("X1") and negative ("X0") classes equal total influence."""
    resp = pd.Series(resp)
    counts = resp.value_counts()
    return (0.5 / resp.map(counts)).to_numpy()


##### Train-test splitting #####

def make_folds_protstrat(data, k=5, seed=123):
    """Train-test split stratified by response class and by protein (so no protein appears
    in both the train and test set of a fold). `data` must have a "response" column with
    values "X0"/"X1", and an index formatted "<uniprot>_<site>".

    Uses sklearn's StratifiedKFold on protein-level labels in place of R's
    caret::createFolds() - both target balanced class proportions per fold, but the exact
    fold membership won't be identical between the two implementations even with the same
    seed, since the algorithms and RNG streams differ.
    """
    uniprot = pd.Series(data.index, index=data.index).str.replace(r"_.+$", "", regex=True)
    protein_resp = (
        pd.DataFrame({"uniprot": uniprot, "response": data["response"].to_numpy()})
        .groupby("uniprot")["response"]
        .apply(lambda x: bool((x == "X1").any()))
    )

    proteins = protein_resp.index.to_numpy()
    labels = protein_resp.to_numpy()
    skf = StratifiedKFold(n_splits=k, shuffle=True, random_state=seed)

    splits = {}
    for i, (train_idx, test_idx) in enumerate(skf.split(proteins, labels), start=1):
        train_proteins = set(proteins[train_idx])
        test_proteins = set(proteins[test_idx])
        splits[f"Fold{i}"] = {
            "train_dat": data[uniprot.isin(train_proteins)],
            "test_dat": data[uniprot.isin(test_proteins)],
        }
    return splits


def make_repeated_protstrat_splits(data, num_reps=1, k=5, seed=123):
    """Repeat make_folds_protstrat() num_reps times with a distinct seed each time, giving
    every fold across every repeat a unique name (e.g. "Rep1Fold1", "Rep2Fold1", ...)."""
    all_splits = {}
    for j in range(1, num_reps + 1):
        for name, split in make_folds_protstrat(data, k=k, seed=seed + j).items():
            all_splits[f"Rep{j}{name}"] = split
    return all_splits


##### Model training #####

def train_logreg_weighted(train_dat, seed=None):
    """Fits an unregularised, balanced-class-weighted logistic regression - see the module
    docstring for how this maps onto R's caret::train(..., method="glm", family="binomial").
    `seed` is accepted for API parity with the R version; LBFGS is deterministic, so it has
    no effect here.
    """
    X = train_dat.drop(columns="response")
    y = (train_dat["response"] == "X1").astype(int)
    weights = case_weights_balanced(train_dat["response"])
    # C=inf (rather than the deprecated penalty=None) switches off L2 regularisation:
    # as C -> inf the penalty term's contribution to the objective vanishes, leaving plain
    # weighted MLE.
    model = LogisticRegression(C=float("inf"), max_iter=1000, random_state=seed)
    model.fit(X, y, sample_weight=weights)
    return model


##### Predictions and results #####

def make_preds_from_models(models, train_test_splits, dat_feats, return_all_preds_lst=False):
    """Applies trained models to their held-out test sets and to `dat_feats`.
    Returns a dict with:
      test_preds_df - test-set predictions, one row per (split, held-out site)
      auc_df        - AUC within each split's test set
      all_preds_df  - median predicted score per site in dat_feats, across all splits
    """
    results = {}

    test_preds = []
    for fold, model in models.items():
        test_dat = train_test_splits[fold]["test_dat"]
        # Reindex to model.feature_names_in_ rather than trusting column order: unlike R's
        # formula interface (which matches by name), sklearn only warns - it does not
        # realign - if predict-time columns are ordered differently than at fit time, which
        # would otherwise silently score every prediction against the wrong features.
        X_test = test_dat.drop(columns="response")[model.feature_names_in_]
        test_preds.append(pd.DataFrame({
            "X1": model.predict_proba(X_test)[:, 1],
            "obs": test_dat["response"].to_numpy(),
            "fold": fold,
            "uniprot_site": test_dat.index,
        }))
    test_preds_df = pd.concat(test_preds, ignore_index=True)
    test_preds_df["Resample"] = test_preds_df["fold"]
    results["test_preds_df"] = test_preds_df

    aucs = {
        fold: roc_auc_score((d["obs"] == "X1").astype(int), d["X1"])
        for fold, d in test_preds_df.groupby("Resample")
    }
    results["auc_df"] = pd.DataFrame({"Resample": list(aucs.keys()), "AUC": list(aucs.values())})

    all_preds = []
    for fold, model in models.items():
        all_preds.append(pd.DataFrame({
            "X1": model.predict_proba(dat_feats[model.feature_names_in_])[:, 1],
            "uniprot_site": dat_feats.index,
            "fold": fold,
        }))
    all_preds_df = pd.concat(all_preds, ignore_index=True)
    results["all_preds_df"] = all_preds_df.groupby("uniprot_site")["X1"].median().reset_index()

    if return_all_preds_lst:
        results["all_preds_lst"] = all_preds
    return results


def train_and_make_preds_logreg(train_test_splits, dat_feats, seed=80, return_all_preds_lst=False):
    """Trains one weighted logistic regression model per train-test split, then scores
    every model against its own test set and against `dat_feats`.
    Returns a dict: auc_df, test_preds_df, all_preds_df, models
    """
    models = {
        fold: train_logreg_weighted(split["train_dat"], seed=seed + j)
        for j, (fold, split) in enumerate(train_test_splits.items(), start=1)
    }

    results = make_preds_from_models(
        models, train_test_splits, dat_feats, return_all_preds_lst=return_all_preds_lst
    )
    results["models"] = models
    return results

File: scripts/model_training.R
##### Background #####
# Core engine for training the weighted logistic regression classifier:
# evaluation metrics, protein-stratified train-test splitting, model training,
# and prediction extraction. See pipeline.R for the user-facing entry point.

library(dplyr)
library(purrr)
library(caret)
library(pROC)
library(yardstick)


##### Evaluation metrics #####

# summaryFunction for caret::trainControl(). Adds ROC AUC (positive class = lev[1])
# to the default Accuracy/Sensitivity/Specificity metrics.
accROCSummary <- function(data, lev = NULL, model = NULL){
  if (is.character(data$obs)) {
    data$obs <- factor(data$obs, levels = lev)
  }
  acc <- caret::postResample(data[, "pred"], data[, "obs"])

  if (length(lev) > 2) {
    stop(paste("Your outcome has", length(lev), "levels. accROCSummary() only supports two-class outcomes."))
  }
  if (!all(levels(data[, "pred"]) == lev)) {
    stop("levels of observed and predicted data do not match")
  }
  rocObject <- try(pROC::roc(data$obs, data[, lev[1]], direction = ">", quiet = TRUE), silent = TRUE)
  rocAUC <- if (inherits(rocObject, "try-error")) NA else rocObject$auc

  out <- c(rocAUC,
           caret::sensitivity(data[, "pred"], data[, "obs"], lev[1]),
           caret::specificity(data[, "pred"], data[, "obs"], lev[2]))
  names(out) <- c("ROC", "Sens", "Spec")
  c(acc, out)
}

# ROC AUC from a data.frame with an observed-class column (factor, levels X0/X1) and a
# predicted-probability-of-X1 column.
ROCAUC_fromPredictor <- function(data, obsCol, predCol){
  data[, c("obs", "pred")] <- data[, c(obsCol, predCol)]
  if (typeof(data$obs) == "logical") {
    data$obs <- factor(make.names(as.numeric(data$obs)), levels = c("X1", "X0"))
  }
  AUCDat <- yardstick::roc_auc(data, truth = "obs", "pred")
  as.numeric(AUCDat[1, ".estimate"])
}


##### Feature/response prep #####

combine_FeatResp <- function(feats, resp){
  comb <- cbind(resp, feats)
  colnames(comb)[1] <- "response"
  comb
}

# Weights that give the positive and negative classes equal total influence
caseWeights_balanced <- function(resp){
  tab <- table(resp)
  unname(0.5 / tab[as.character(resp)])
}


##### Train-test splitting #####

# 5-fold cross-validation control used inside model training, scored with accROCSummary
fitControl_CV <- caret::trainControl(
  method = "cv", number = 5,
  classProbs = TRUE,
  savePredictions = TRUE,
  summaryFunction = accROCSummary
)

# Train-test split stratified by response class and by protein (so no protein appears in
# both the train and test set of a fold). `data` must have a "response" column with values
# "X0"/"X1", and rownames formatted "<uniprot>_<site>".
makeFoldsProtstrat <- function(seed = 123, k = 5, data){
  data$uniprot <- gsub("_.+$", "", rownames(data))

  respMax <- function(x) "X1" %in% x
  protResp <- data[, c("uniprot", "response")] %>%
    group_by(uniprot) %>%
    summarise(response = respMax(response))

  set.seed(seed)
  protFolds <- caret::createFolds(protResp$response, k = k, returnTrain = TRUE)

  splits <- lapply(protFolds, function(trainIdx){
    trainProt <- protResp$uniprot[trainIdx]
    testProt  <- protResp$uniprot[-trainIdx]
    list(
      trainDat = subset(data, uniprot %in% trainProt) %>% dplyr::select(-uniprot),
      testDat  = subset(data, uniprot %in% testProt) %>% dplyr::select(-uniprot)
    )
  })
  names(splits) <- paste0("Fold", seq_along(splits))
  splits
}

# Repeat makeFoldsProtstrat() numReps times with a distinct seed each time, giving every
# fold across every repeat a unique name (e.g. "Rep1Fold1", "Rep2Fold1", ...).
makeRepeatedProtstratSplits <- function(data, numReps = 1, k = 5, seed = 123){
  allSplits <- list()
  for (j in seq_len(numReps)) {
    splits <- makeFoldsProtstrat(seed = seed + j, k = k, data = data)
    names(splits) <- paste0("Rep", j, names(splits))
    allSplits <- c(allSplits, splits)
  }
  allSplits
}


##### Model training #####

modTrain_LogReg_weighted <- function(feats, resp, fitControl, seed,
                                     weightFunc = caseWeights_balanced){
  weights <- weightFunc(resp)
  FeatResp_data <- combine_FeatResp(feats = feats, resp = resp)
  set.seed(seed)
  caret::train(response ~ ., data = FeatResp_data,
               method = "glm", family = "binomial",
               metric = "Accuracy", trControl = fitControl,
               weights = weights)
}


##### Predictions and results #####

# Applies trained models to their held-out test sets and to `datFeats`.
# Returns:
#   modsPredsTest_df - test-set predictions, one row per (split, held-out site)
#   AUCdf            - AUC within each split's test set
#   allPredsDf       - median predicted score per site in datFeats, across all splits
makePreds_fromMods <- function(mods, trainTestSplits, datFeats, returnAllPredsLst = FALSE){
  resultsLst <- list()

  modsPredsTest <- map2(mods, names(mods), function(mod, fold){
    testDat <- trainTestSplits[[fold]]$testDat
    predictions <- predict(mod, newdata = testDat[, -which(colnames(testDat) == "response")], type = "prob")
    data.frame(X1 = predictions[, "X1"], obs = testDat$response,
               fold = fold, uniprot_site = rownames(testDat))
  })
  modsPredsTest_df <- purrr::reduce(modsPredsTest, rbind) %>% mutate(Resample = fold)
  resultsLst$modsPredsTest_df <- modsPredsTest_df

  AUCs <- modsPredsTest_df %>% split(.$Resample) %>%
    map(~ROCAUC_fromPredictor(data = ., obsCol = "obs", predCol = "X1")) %>% unlist
  resultsLst$AUCdf <- data.frame(Resample = names(AUCs), AUC = AUCs)

  allPredsLst <- map2(mods, names(mods), function(mod, fold){
    predictions <- predict(mod, newdata = datFeats, type = "prob")
    data.frame(X1 = predictions[, "X1"], uniprot_site = rownames(datFeats))
  })
  allPredsLstDf <- map_dfr(allPredsLst, function(x) x, .id = "fold")
  resultsLst$allPredsDf <- allPredsLstDf %>%
    group_by(uniprot_site) %>%
    summarise(X1 = median(X1, na.rm = TRUE))

  if (returnAllPredsLst) resultsLst$allPredsLst <- allPredsLst
  resultsLst
}

# Trains one weighted logistic regression model per train-test split, then scores every
# model against its own test set and against `datFeats`.
# Returns a list: AUCdf, modsPredsTest_df, allPredsDf, mods
trainAndMakePreds_logreg <- function(trainTestSplits, datFeats, seed = 80, returnAllPredsLst = FALSE){
  mods <- vector("list", length(trainTestSplits))
  for (j in seq_along(trainTestSplits)) {
    trainDat <- trainTestSplits[[j]]$trainDat
    mods[[j]] <- modTrain_LogReg_weighted(
      feats = trainDat[, -which(colnames(trainDat) == "response")],
      resp = trainDat$response,
      fitControl = fitControl_CV, seed = seed + j,
      weightFunc = caseWeights_balanced
    )
  }
  names(mods) <- names(trainTestSplits)

  resultsLst <- makePreds_fromMods(mods = mods, trainTestSplits = trainTestSplits,
                                   datFeats = datFeats, returnAllPredsLst = returnAllPredsLst)
  resultsLst$mods <- mods
  resultsLst
}
